0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,218 @@
+| ... | _(Optional)_ Additional arbitrary headers. Where appropriate, it is RECOMMENDED to match the headers defined for HTTP messages ([RFC7230](https://tools.ietf.org/html/rfc7230#section-3.2)) |
+
+# Pseudocode
+
+```js
+/** Executed by receiver at protocol setup
+  * @param account (this is the address the receiver will listen on, as
+  * returned by their ledger plugin)
+  */ 
+
+// The token is attached to the receiver's ILP address, allowing them to
+// regenerate the shared secret every time they receive a payment, instead of
+// keeping a store of all shared secrets they've given out.
+
+receiver_secret = random_bytes(32)
+token = random_bytes(16)
+
+// The receiver id differentiates this receiver from other receivers on the
+// same account with different receiver_secrets. When they receive a payment,
+// the receiver can make sure the receiver id in the ILP address is their own.
+
+receiver_id = hmac_sha_256(receiver_secret, "ilp_psk_receiver_id").slice(0, 8)
+receiver_address = account + "." + base64url(receiver_id) + base64url(token)
+
+// This pattern of taking the HMAC of a secret and a hard-coded string will be
+// repeated for the generation of many variables. It helps secure the secrets
+// against any interactions between the algorithms used.
+
+shared_secret_generator = hmac_sha_256(receiver_secret, "ilp_psk_generation")
+shared_secret = hmac_sha_256(shared_secret_generator, token)
+
+// Now the shared_secret and the receiver_address are transmitted to the
+// sender.
+
+/** Executed by sender before each payment, to get the ILP packet and the condition.
+  * @param shared_secret (sent by the receiver)
+  * @param receiver_address (sent by the receiver)
+  *
+  * @param headers (a list of HTTP style headers, which will be encrypted)
+  * @param unsafe_headers (a list of HTTP style headers, which will be left
+  * unencrypted)
+  * @param application_data (arbitrary binary data used for higher level
+  * protocols, which will be encrypted)
+  * @param destination_amount (amount that the receiver will receive on their
+  * ledger, used to construct the ILP packet)
+  */
+
+// The private headers are put together first
+
+private_headers = create_empty_buffer()
+private_headers += headers.join('\n')
+
+// Application data is attached to the private headers, after an empty line.
+
+private_headers += '\n\n'
+private_headers += application_data
+
+// A key is created in order to encrypt the private headers and the application
+// data. A nonce is used to generate the key, so that the same private headers
+// won't be turned into the same ciphertext.
+
+nonce = random_bytes(16)
+payment_encryption_key = hmac_sha_256(shared_secret, nonce)
+encrypted_data = aes_256_ctr(payment_encryption_key, private_headers)
+
+// Now the PSK data object is constructed, containing the public headers, the
+// encrypted private headers, and the encrypted application data. A status
```

Comment:
The second sentence here is unnecessary because it just says exactly what the code below it does